src/program/latex/mod.rs

```diff
@@ -0,0 +1,1294 @@
+\usepackage[margin=1in]{geometry}
+\usepackage{tikz}
+\usetikzlibrary{quantikz}
+\begin{document}
+\begin{tikzcd}"
+                .to_string(),
+            body: String::new(),
+            footer: r"\end{tikzcd}
+\end{document}"
+                .to_string(),
+        }
+    }
+}
+
+impl Display for Document {
+    /// Returns the entire document in LaTeX string.
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}{}{}", self.header, self.body, self.footer)
+    }
+}
+
+/// A Diagram represents a collection of wires in a circuit. It encodes the
+/// wires row in the diagram and its relationship to other wires. A row is one
+/// of the wires in the circuit BTreeMap. Diagram tracks relationships between
+/// wires with two pieces of information--1. the wires row (its order in the
+/// BTreeMap), and 2. the column that spreads between all wires that pass
+/// through a multi qubit gate, e.g. CNOT. The size of the diagram can be
+/// measured by multiplying the column with the length of the circuit. This is
+/// an [m x n] matrix where each element in the matrix represents an item to be
+/// rendered onto the diagram using one of the Quantikz commands.
+#[derive(Clone, Debug, Default)]
+struct Diagram {
+    /// customizes how the diagram renders the circuit
+    settings: RenderSettings,
+    /// total number of elements on each wire
+    column: u32,
+    /// column at which qubits in positional order form relationships
+    relationships: HashMap<u32, Vec<Qubit>>,
+    /// a BTreeMap of wires with the name of the wire as the key
+    circuit: BTreeMap<u64, Box<Wire>>,
+}
+
+impl Diagram {
+    /// Compares qubits from a single instruction associated with a column on
+    /// the circuit to all of the qubits used in the quil program. If a qubit
+    /// from the quil program is not found in the qubits in the single
+    /// instruction line, then an empty slot is added to that column on the
+    /// qubit wire of the circuit indicating a "do nothing" at that column.
+    ///
+    /// # Arguments
+    /// `&mut self` - exposes the wires on the Circuit
+    /// `qubits` - exposes the qubits used in the Program
+    /// `instruction` - exposes the qubits in a single Instruction
+    fn set_empty(&mut self, qubits: &HashSet<Qubit>, instruction: &Instruction) {
+        'program_loop: for program_qubit in qubits {
+            if let Instruction::Gate(gate) = instruction {
+                for gate_qubit in &gate.qubits {
+                    if program_qubit == gate_qubit {
+                        continue 'program_loop;
+                    }
+                }
+
+                if let Qubit::Fixed(q) = program_qubit {
+                    if let Some(wire) = self.circuit.get_mut(q) {
+                        wire.empty.insert(self.column, Command::Qw);
+                    }
+                }
+            }
+        }
```

Comment:
I think we can restructure this to only do something if `instruction` is a `Gate` right? YAlso, this was a fun one to play iterator golf on:

```suggestion
        if let Instruction::Gate(gate) = instruction {
            qubits
                .difference(&gate.qubits.iter().cloned().collect())
                .filter_map(|q| match q {
                    Qubit::Fixed(index) => Some(index),
                    _ => None,
                })
                .for_each(|index| {
                    self.circuit
                        .get_mut(index)
                        .map(|wire| wire.empty.insert(self.column, Command::Qw));
                });
        }

```